pyiceberg/table/__init__.py

```diff
@@ -533,6 +551,39 @@ def _(update: SetCurrentSchemaUpdate, base_metadata: TableMetadata, context: _Ta
     return base_metadata.model_copy(update={"current_schema_id": new_schema_id})
 
 
+@_apply_table_update.register(AddPartitionSpecUpdate)
+def _(update: AddPartitionSpecUpdate, base_metadata: TableMetadata, context: _TableMetadataUpdateContext) -> TableMetadata:
+    for spec in base_metadata.partition_specs:
+        if spec.spec_id == update.spec_id:
+            raise ValueError(f"Partition spec with id {spec.spec_id} already exists: {spec}")
+
+    context.add_update(update)
+    return base_metadata.model_copy(
+        update={
+            "partition_specs": base_metadata.partition_specs + [update.spec],
+        }
+    )
+
+
+@_apply_table_update.register(SetDefaultSpecUpdate)
+def _(update: SetDefaultSpecUpdate, base_metadata: TableMetadata, context: _TableMetadataUpdateContext) -> TableMetadata:
+    new_spec_id = update.spec_id
+    if new_spec_id == base_metadata.default_spec_id:
```

Comment:
See here https://github.com/apache/iceberg-python/pull/245/files#r1475565567

Reply:
Thanks for the update! I think we also need to add some logic here like:
https://github.com/apache/iceberg-python/blob/7fbcc220263228618da8d6871a50a0b82e22a843/pyiceberg/table/__init__.py#L537-L541
to make other catalogs such as `hive` to handle the `-1` correctly. WDYT?

Java ref: https://github.com/apache/iceberg/blob/9921937d8285dec9a19fd16b0cd82d451a8aca9e/core/src/main/java/org/apache/iceberg/TableMetadata.java#L1079-L1084

Reply:
Ah interesting,  I didn't know we actually supported this -1 for non-rest catalogs. I thought setting the schema ID as -1 was purely spec'd in the REST Catalog. Since the table metadata API in java allows it, we probably should also support it here.

Reply:
Updated